### PR TITLE
fix(gnovm): change initialization order so realm objects are persisted before referenced by other realms

### LINKED ISSUE #974: Passing a pointer to another realm that assigns it allows to steal objects ownership
### Description

If you pass a pointer to an object `x` in realm `A` to a method of a realm `B` that stores this pointer, the ownership is set to `B`

### Your environment

* macos 13.4.1
* go version go1.20.4 darwin/arm64
* gno master 52ea535634961f87e90dd080bc3aa22986ac0142

### Steps to reproduce

- Deploy these two realms:

```go
package steal_ownership

var ptr *uint32

func Steal(xptr *uint32) {
	ptr = xptr
}
```

```go
package mis_ownership

import "gno.land/r/demo/bugs/steal_ownership"

var (
	x   = uint32(42)
	ptr = &x
)

func init() {
	steal_ownership.Steal(ptr)
}

func MutateDeref() {
	*ptr = 21
}

func MutatePtr() {
	y := uint32(21)
	ptr = &y
}

func MutateValue() {
	x = 21
}
```

- Call `MutateDeref` or `MutatePtr` or `MutateValue` on `mis_ownership` realm

### Expected behaviour

Mutations run correctly

### Actual behaviour

Mutations error out

```
--= Error =--
Data: cannot modify external-realm or non-realm object: object pkg path: gno.land/r/demo/bugs/steal_ownership, realm: gno.land/r/demo/bugs/mis_ownership
```

### Additional info
You can find the full code here https://github.com/TERITORI/gno/pull/4

I discovered it while passing interfaces in this way so this is not limited to raw pointers

I edited the gnovm code to print realm package path in the error, you can see the code here https://github.com/TERITORI/gno/pull/4/files#diff-7041eca77b078f22ee334239995c403673f040b3726bcffa5a8a053b68a951a7

### PR DESCRIPTION
**Pull request by deelawn**
Closes #974.

The first part of this PR addresses the issue linked above.  The issue details a scenario in which pointer ownership was able to be stolen by passing the pointer value to another realm that saved the value to it's own realm during initialization. The solution for this is to persist the mem package _before_ running any `init` functions. This should be fine because realm finalization happens the same number of times.

The realm initialization behavior before (still the behavior for packages):
1. Finalize realm after each `init` function
2. Finalize the mem package

New behavior for realms:
1. Finalize the mem package
2. Finalize the realm after each `init` function
---

~~The second part of this PR attempts to clarify what the expected behavior should be when sharing pointers to values amongst realms. It achieves this by updating the effective gno docs to detail the expected behavior and to generally discourage users from doing things like this unless there is a good reason to do so and they have a strong understanding of how this is expected to work.  There are also clarifying txtar tests that have been added to test the expected behavior.~~

TODO: remove the above docs changes. It was decided not to allow realms to store references to objects in other realms -- #2332  
<details><summary>Contributors' checklist...</summary>

- [x] Added new tests, or not needed, or not feasible
- [x] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [x] Updated the official documentation or not needed
- [x] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [x] Added references to related issues and PRs
- [x] Provided any useful hints for running manual tests
- [x] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

### CODE REVIEW

**Comment by deelawn** [comment]
Marking as draft until the failing tests are fixed.  I took a look and I see two categories of failing tests:
1. Tests are failing that are relying on state to be saved in packages. Tests previously allowed package state because the mem package was persisted after running init.  Now that the mem package is saved before running init, state variables initialized within the init function won't get persisted (maybe)
2. Saving the mem package before init is causing an issue when trying to assign a value to a realm interface type variable from within the init.

File: gnovm/pkg/gnolang/machine.go [hunk 1]
```diff
@@ -481,10 +480,16 @@ func (m *Machine) injectLocOnPanic() {
 // Add files to the package's *FileSet and run them.
 // This will also run each init function encountered.
 func (m *Machine) RunFiles(fns ...*FileNode) {
-	m.runFiles(fns...)
+	m.runFiles(nil, fns...)
+}
+
+// RunFilesWithMemPkg is almost the same as RunFiles; the difference is that it
+// saves the mempackage to the store.
```
Comment (jaekwon) [inline gnovm/pkg/gnolang/machine.go:487]: the name is inconsistent with its behavior though...

Comment (jaekwon) [reply gnovm/pkg/gnolang/machine.go:487]: gut tells me not to add saving logic into runFiles logic, but instead runFiles could return new init functions as []Name, and the caller can save before or after as needed.

I'mstill trying to wrap my head around why it's needed. thinking..

**Comment by zivkovicmilos** [comment]
Relevant, older PR by @n0izn0iz :
https://github.com/gnolang/gno/pull/1074

**Comment by jaekwon** [comment]
back on this.

**Comment by Kouteki** [comment]
@mvertes @petar-dambovaliev please look into this

**Comment by deelawn** [comment]
Closing; #2255 also fixed this issue.

**Pull request closed without merging**